File: src/main/core/dxvk.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const zlib = require('node:zlib');
const { downloadFile, fetchJson } = require('./download');

/**
 * DXVK в один клик (3.1).
 *
 * DXVK переводит DirectX 8–11 на Vulkan. В старых играх это часто убирает
 * фризы и поднимает FPS. Ставится он просто — нужные DLL кладутся рядом
 * с exe игры, — но руками в этом легко ошибиться: взять 32-битные файлы
 * для 64-битной игры, положить dxgi.dll в игру на DirectX 9 или затереть
 * собственную DLL игры. Здесь всё это делается само:
 *
 *   1. По заголовку exe узнаём разрядность, по таблице импорта — DirectX.
 *      Движок часто грузит DirectX не из exe, а из своей DLL рядом
 *      (UnityPlayer.dll, engine.dll) — смотрим и их.
 *   2. Скачиваем свежий DXVK с GitHub (проверяем sha256, который дал GitHub)
 *      и берём из архива только нужные DLL нужной разрядности.
 *   3. Если в папке уже лежит файл с тем же именем, он уходит в резервную
 *      копию и возвращается при удалении DXVK.
 *   4. В папке игры остаётся отметка dxvk.modlaunch.json: что поставили
 *      и что сохранили — по ней DXVK убирается без следов.
 *
 * ReShade для DirectX 10–11 сам лежит как dxgi.dll, и вместе с DXVK так
 * не работает — в этом случае честно говорим об этом и ничего не трогаем.
 */

const RELEASES = 'https://api.github.com/repos/doitsujin/dxvk/releases/latest';
const MARKER = 'dxvk.modlaunch.json';
const BACKUP_SUFFIX = '.modlaunch-backup';

/** Какие DLL нужны для какого DirectX. */
const FILES = {
  dx8: ['d3d8.dll', 'd3d9.dll'],
  dx9: ['d3d9.dll'],
  dx10: ['d3d10core.dll', 'd3d11.dll', 'dxgi.dll'],
  dx11: ['d3d10core.dll', 'd3d11.dll', 'dxgi.dll'],
};
const APIS = Object.keys(FILES);

/* --- PE: разрядность и импорты ---------------------------------------- */

function readAt(fd, offset, length) {
  const buffer = Buffer.alloc(length);
  const read = fs.readSync(fd, buffer, 0, length, offset);
  return read === length ? buffer : buffer.subarray(0, read);
}

/**
 * Разрядность и список импортируемых DLL у exe или dll.
 * Читаем только заголовки и таблицу импорта, а не весь файл.
 * @returns {{arch: 'x86'|'x64', imports: string[]}|null}
 */
function peInfo(file) {
  let fd;
  try {
    fd = fs.openSync(file, 'r');
    const dos = readAt(fd, 0, 64);
    if (dos.length < 64 || dos.toString('latin1', 0, 2) !== 'MZ') return null;
    const peOffset = dos.readUInt32LE(0x3c);
    const head = readAt(fd, peOffset, 24);
    if (head.length < 24 || head.toString('latin1', 0, 4) !== 'PE\0\0') return null;
    const machine = head.readUInt16LE(4);
    const sections = head.readUInt16LE(6);
    const optSize = head.readUInt16LE(20);
    const opt = readAt(fd, peOffset + 24, optSize);
    const magic = opt.readUInt16LE(0);
    const arch = machine === 0x8664 || magic === 0x20b ? 'x64' : machine === 0x14c ? 'x86' : null;
    if (!arch) return null;

    // Каталог данных: у PE32 с 96-го байта, у PE32+ — со 112-го. Импорт — второй.
    const dirs = magic === 0x20b ? 112 : 96;
    const importRva = opt.length >= dirs + 16 ? opt.readUInt32LE(dirs + 8) : 0;
    const table = readAt(fd, peOffset + 24 + optSize, sections * 40);
    const toOffset = (rva) => {
      for (let i = 0; i < sections; i++) {
        const s = i * 40;
        const va = table.readUInt32LE(s + 12);
        const size = Math.max(table.readUInt32LE(s + 8), table.readUInt32LE(s + 16));
        if (rva >= va && rva < va + size) return rva - va + table.readUInt32LE(s + 20);
      }
      return null;
    };

    const imports = [];
    const at = importRva ? toOffset(importRva) : null;
    if (at != null) {
      for (let i = 0; i < 512; i++) {
        const desc = readAt(fd, at + i * 20, 20);
        if (desc.length < 20 || desc.every((b) => b === 0)) break;
        const nameAt = toOffset(desc.readUInt32LE(12));
        if (nameAt == null) continue;
        const raw = readAt(fd, nameAt, 64).toString('latin1');
        const name = raw.slice(0, raw.indexOf('\0') === -1 ? raw.length : raw.indexOf('\0')).toLowerCase();
        if (name) imports.push(name);
      }
    }
    return { arch, imports };
  } catch {
    return null;
  } finally {
    if (fd !== undefined) fs.closeSync(fd);
  }
}

/** DirectX по списку импортов. null — не поняли (или это не DirectX 8–11). */
function apiFromImports(imports) {
  const has = (name) => imports.includes(name);
  if (has('d3d12.dll')) return 'dx12';
  if (has('d3d11.dll') || has('dxgi.dll')) return 'dx11';
  if (has('d3d10.dll') || has('d3d10_1.dll') || has('d3d10core.dll')) return 'dx10';
  if (has('d3d9.dll')) return 'dx9';
  if (has('d3d8.dll')) return 'dx8';
  if (has('opengl32.dll')) return 'opengl';
  if (has('vulkan-1.dll')) return 'vulkan';
  return null;
}

/**
 * Что за игра: разрядность exe и её DirectX.
 * @returns {{exe: string, dir: string, name: string, arch: string, api: string|null, supported: boolean}}
 */
function inspect(exe) {
  const info = peInfo(exe);
  if (!info) throw Object.assign(new Error('not an exe'), { code: 'DXVK_NOT_EXE' });
  const dir = path.dirname(exe);
  let api = apiFromImports(info.imports);

  // Движок нередко грузит DirectX из своей DLL рядом с exe.
  if (!api || api === 'opengl') {
    let entries = [];
    try {
      entries = fs.readdirSync(dir).filter((n) => /\.dll$/i.test(n)).slice(0, 80);
    } catch {
      /* нет доступа к папке — останемся с тем, что знаем */
    }
    for (const name of entries) {
      if (FILES.dx11.includes(name.toLowerCase()) || /^d3d[89]\.dll$/i.test(name)) continue;
      const found = apiFromImports(peInfo(path.join(dir, name))?.imports ?? []);
      if (found && found !== 'opengl' && found !== 'vulkan') {
        api = found;
        break;
      }
      if (found && !api) api = found;
    }
  }

  return {
    exe,
    dir,
    name: gameName(exe),
    arch: info.arch,
    api,
    supported: APIS.includes(api),
  };
}

/** Имя игры для списка: папка, где лежит exe (без «bin», «x64» и т. п.). */
function gameName(exe) {
  let dir = path.dirname(exe);
  while (/^(bin(32|64)?|x64|x86|win(32|64)|binaries|release|game)$/i.test(path.basename(dir))) {
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return path.basename(dir) || path.basename(exe, path.extname(exe));
}

/* --- tar.gz ----------------------------------------------------------- */

/**
 * Файлы из tar.gz (DXVK выпускается так). Нужны только обычные файлы;
 * длинные имена бывают в записях GNU «L» и pax «x» — их тоже понимаем.
 * @returns {Map<string, Buffer>} путь → содержимое
 */
function readTarGz(buffer) {
  const tar = zlib.gunzipSync(buffer);
  const out = new Map();
  let offset = 0;
  let longName = null;
  while (offset + 512 <= tar.length) {
    const header = tar.subarray(offset, offset + 512);
    if (header.every((b) => b === 0)) break;
    const field = (start, length) => {
      const raw = header.toString('utf8', start, start + length);
      const end = raw.indexOf('\0');
      return end === -1 ? raw : raw.slice(0, end);
    };
    const size = parseInt(field(124, 12).trim() || '0', 8) || 0;
    const type = field(156, 1) || '0';
    const prefix = field(345, 155);
    let name = longName ?? (prefix ? `${prefix}/${field(0, 100)}` : field(0, 100));
    longName = null;
    const body = tar.subarray(offset + 512, offset + 512 + size);
    if (type === 'L') longName = body.toString('utf8').replace(/\0+$/, '');
    else if (type === 'x') {
      const pathLine = /\d+ path=([^\n]+)\n/.exec(body.toString('utf8'));
      if (pathLine) longName = pathLine[1];
    } else if (type === '0' || type === '\0') out.set(name.replace(/^\.\//, ''), Buffer.from(body));
    offset += 512 + Math.ceil(size / 512) * 512;
  }
  return out;
}

/** Это ReShade под именем системной DLL (dxgi.dll, d3d9.dll…)? */
function isReShade(file) {
  try {
    if (!fs.statSync(file).isFile()) return false;
    const text = fs.readFileSync(file).toString('latin1');
    return text.includes('crosire') || text.includes('ReShade');
  } catch {
    return false;
  }
}

/* --- установка -------------------------------------------------------- */

class Dxvk {
  /**
   * @param {object} [options]
   * @param {string} [options.cacheDir] где держать скачанный архив
   * @param {Function} [options.download]
   * @param {Function} [options.getJson]
   */
  constructor({ cacheDir, download = downloadFile, getJson = fetchJson } = {}) {
    this.cacheDir = cacheDir;
    this.download = download;
    this.getJson = getJson;
  }

  /** Стоит ли DXVK в этой папке (по нашей отметке). */
  status(dir) {
    try {
      const data = JSON.parse(fs.readFileSync(path.join(dir, MARKER), 'utf8'));
      return { installed: true, version: data.version ?? '', arch: data.arch, api: data.api, files: data.files ?? [], installedAt: data.installedAt ?? null };
    } catch {
      return { installed: false };
    }
  }

  /** Свежий выпуск DXVK: версия, ссылка, sha256. */
  async latest() {
    const data = await this.getJson(RELEASES, { timeout: 20000, headers: { Accept: 'application/vnd.github+json' } });
    const asset = (data?.assets ?? []).find((a) => /^dxvk-[\d.]+\.tar\.gz$/i.test(a?.name ?? ''));
    if (!asset) throw Object.assign(new Error('no dxvk asset'), { code: 'DXVK_DOWNLOAD' });
    const sha256 = /^sha256:([0-9a-f]{64})$/i.exec(String(asset.digest ?? ''))?.[1] ?? undefined;
    return { version: String(data.tag_name ?? '').replace(/^v/i, ''), url: asset.browser_download_url, name: asset.name, sha256 };
  }

  /** Архив DXVK: из кэша, если эта версия уже скачана. */
  async archive(onStage = () => {}) {
    const release = await this.latest();
    const cached = this.cacheDir ? path.join(this.cacheDir, release.name) : null;
    if (cached && fs.existsSync(cached)) return { release, file: cached };
    onStage({ stage: 'download', ratio: 0 });
    const file = await this.download(release.url, {
      fileName: release.name,
      sha256: release.sha256,
      onProgress: (p) => onStage({ stage: 'download', ratio: p.ratio }),
    });
    if (!cached) return { release, file };
    fs.mkdirSync(this.cacheDir, { recursive: true });
    fs.copyFileSync(file, cached);
    fs.rmSync(file, { force: true });
    return { release, file: cached };
  }

  /**
   * Поставить DXVK рядом с exe.
   * @param {string} exe
   * @param {{api?: string}} [options] DirectX, если его не удалось определить самим
   */
  async install(exe, { api: chosen } = {}, onStage = () => {}) {
    const game = inspect(exe);
    const api = APIS.includes(chosen) ? chosen : game.api;
    if (!APIS.includes(api)) throw Object.assign(new Error(`unsupported api ${api}`), { code: 'DXVK_API', api: game.api });
    if (this.status(game.dir).installed) this.remove(game.dir);

    const names = FILES[api];
    if (names.some((n) => isReShade(path.join(game.dir, n)))) {
      throw Object.assign(new Error('reshade conflict'), { code: 'DXVK_RESHADE' });
    }

    const { release, file } = await this.archive(onStage);
    onStage({ stage: 'install' });
    const entries = readTarGz(fs.readFileSync(file));
    const folder = game.arch === 'x64' ? 'x64' : 'x32';
    const pick = (name) => [...entries.entries()].find(([p]) => p.toLowerCase().endsWith(`/${folder}/${name}`))?.[1];
    const files = names.filter((n) => pick(n));
    if (!files.length) throw Object.assign(new Error('dxvk archive has no dlls'), { code: 'DXVK_DOWNLOAD' });

    const backups = [];
    const written = [];
    try {
      for (const name of files) {
        const target = path.join(game.dir, name);
        if (fs.existsSync(target)) {
          fs.renameSync(target, target + BACKUP_SUFFIX);
          backups.push(name);
        }
        fs.writeFileSync(target, pick(name));
        written.push(name);
      }
      fs.writeFileSync(
        path.join(game.dir, MARKER),
        JSON.stringify({ version: release.version, arch: game.arch, api, files: written, backups, exe: path.basename(exe), installedAt: new Date().toISOString() }, null, 2)
      );
    } catch (error) {
      // Не оставляем игру наполовину переделанной.
      for (const name of written) fs.rmSync(path.join(game.dir, name), { force: true });
      for (const name of backups) {
        const target = path.join(game.dir, name);
        if (fs.existsSync(target + BACKUP_SUFFIX)) fs.renameSync(target + BACKUP_SUFFIX, target);
      }
      throw error;
    }
    return { ...game, api, version: release.version, files: written, backups };
  }

  /** Убрать DXVK и вернуть то, что лежало на месте его файлов. */
  remove(dir) {
    const marker = path.join(dir, MARKER);
    let data;
    try {
      data = JSON.parse(fs.readFileSync(marker, 'utf8'));
    } catch {
      return { removed: false };
    }
    for (const name of data.files ?? []) {
      if (!FILES.dx8.concat(FILES.dx11).includes(String(name).toLowerCase())) continue;
      fs.rmSync(path.join(dir, name), { force: true });
    }
    for (const name of data.backups ?? []) {
      const target = path.join(dir, path.basename(String(name)));
      if (fs.existsSync(target + BACKUP_SUFFIX)) fs.renameSync(target + BACKUP_SUFFIX, target);
    }
    fs.rmSync(marker, { force: true });
    return { removed: true };
  }
}

module.exports = { Dxvk, peInfo, apiFromImports, inspect, readTarGz, gameName, isReShade, FILES, MARKER };

File: desktop/ModLaunch/Features/Dxvk.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using ModLaunch.Core;
using ModLaunch.Mods;

namespace ModLaunch.Features;

public sealed record ExeInfo(string Exe, string Dir, string Name, string Arch, string? Api, bool Supported);
public sealed record DxvkState(bool Installed, string Version = "", string? Arch = null, string? Api = null);

/// <summary>
/// DXVK для любой игры: читаем exe (разрядность и какой DirectX он грузит),
/// берём свежий выпуск с GitHub и кладём dll рядом, сохранив старые файлы.
/// Метка dxvk.modlaunch.json в папке игры — общая с 3.x.
/// </summary>
public static class Dxvk
{
    const string Releases = "https://api.github.com/repos/doitsujin/dxvk/releases/latest";
    const string Marker = "dxvk.modlaunch.json";
    const string BackupSuffix = ".modlaunch-backup";

    public static readonly Dictionary<string, string[]> Files = new()
    {
        ["dx8"] = ["d3d8.dll", "d3d9.dll"],
        ["dx9"] = ["d3d9.dll"],
        ["dx10"] = ["d3d10core.dll", "d3d11.dll", "dxgi.dll"],
        ["dx11"] = ["d3d10core.dll", "d3d11.dll", "dxgi.dll"],
    };

    // ---------------------------------------------------------------- PE

    public static (string Arch, List<string> Imports)? PeInfo(string file)
    {
        try
        {
            using var fs = File.OpenRead(file);
            using var br = new BinaryReader(fs);
            if (fs.Length < 64 || br.ReadUInt16() != 0x5A4D) return null;
            fs.Position = 0x3C;
            var pe = br.ReadUInt32();
            fs.Position = pe;
            if (br.ReadUInt32() != 0x00004550) return null;
            var machine = br.ReadUInt16();
            var sections = br.ReadUInt16();
            fs.Position = pe + 20;
            var optSize = br.ReadUInt16();
            fs.Position = pe + 24;
            var opt = br.ReadBytes(optSize);
            var magic = BitConverter.ToUInt16(opt, 0);
            var arch = machine == 0x8664 || magic == 0x20b ? "x64" : machine == 0x14c ? "x86" : null;
            if (arch is null) return null;
            var dirs = magic == 0x20b ? 112 : 96;
            var importRva = opt.Length >= dirs + 16 ? BitConverter.ToUInt32(opt, dirs + 8) : 0;
            fs.Position = pe + 24 + optSize;
            var table = br.ReadBytes(sections * 40);
            long? ToOffset(uint rva)
            {
                for (var i = 0; i < sections; i++)
                {
                    var s = i * 40;
                    var va = BitConverter.ToUInt32(table, s + 12);
                    var size = Math.Max(BitConverter.ToUInt32(table, s + 8), BitConverter.ToUInt32(table, s + 16));
                    if (rva >= va && rva < va + size) return rva - va + BitConverter.ToUInt32(table, s + 20);
                }
                return null;
            }
            var imports = new List<string>();
            if (importRva != 0 && ToOffset(importRva) is long at)
            {
                for (var i = 0; i < 512; i++)
                {
                    fs.Position = at + i * 20;
                    var desc = br.ReadBytes(20);
                    if (desc.Length < 20 || desc.All(b => b == 0)) break;
                    if (ToOffset(BitConverter.ToUInt32(desc, 12)) is not long nameAt) continue;
                    fs.Position = nameAt;
                    var raw = Encoding.Latin1.GetString(br.ReadBytes(64));
                    var name = raw.Split('\0')[0].ToLowerInvariant();
                    if (name != "") imports.Add(name);
                }
            }
            return (arch, imports);
        }
        catch { return null; }
    }

    public static string? ApiFromImports(IList<string> imports)
    {
        bool Has(string n) => imports.Contains(n);
        if (Has("d3d12.dll")) return "dx12";
        if (Has("d3d11.dll") || Has("dxgi.dll")) return "dx11";
        if (Has("d3d10.dll") || Has("d3d10_1.dll") || Has("d3d10core.dll")) return "dx10";
        if (Has("d3d9.dll")) return "dx9";
        if (Has("d3d8.dll")) return "dx8";
        if (Has("opengl32.dll")) return "opengl";
        if (Has("vulkan-1.dll")) return "vulkan";
        return null;
    }

    public static ExeInfo Inspect(string exe)
    {
        var info = PeInfo(exe) ?? throw new InvalidDataException(I18n.T("err.DXVK_NOT_EXE"));
        var dir = Path.GetDirectoryName(exe)!;
        var api = ApiFromImports(info.Imports);
        if (api is null or "opengl")
        {
            // Движок часто в отдельной dll рядом (UnityPlayer.dll, engine.dll).
            IEnumerable<string> dlls = [];
            try { dlls = Directory.EnumerateFiles(dir, "*.dll").Take(80).ToList(); } catch { }
            foreach (var dll in dlls)
            {
                var n = Path.GetFileName(dll).ToLowerInvariant();
                if (Files["dx11"].Contains(n) || Regex.IsMatch(n, @"^d3d[89]\.dll$")) continue;
                var found = ApiFromImports(PeInfo(dll)?.Imports ?? []);
                if (found is not null and not "opengl" and not "vulkan") { api = found; break; }
                if (found is not null && api is null) api = found;
            }
        }
        return new ExeInfo(exe, dir, GameName(exe), info.Arch, api, api is not null && Files.ContainsKey(api));
    }

    static string GameName(string exe)
    {
        var dir = Path.GetDirectoryName(exe)!;
        while (Regex.IsMatch(Path.GetFileName(dir), @"^(bin(32|64)?|x64|x86|win(32|64)|binaries|release|game)$", RegexOptions.IgnoreCase))
        {
            var parent = Path.GetDirectoryName(dir);
            if (parent is null || parent == dir) break;
            dir = parent;
        }
        return Path.GetFileName(dir) is { Length: > 0 } name ? name : Path.GetFileNameWithoutExtension(exe);
    }

    // ---------------------------------------------------------------- tar.gz

    public static Dictionary<string, byte[]> ReadTarGz(string file)
    {
        using var gz = new GZipStream(File.OpenRead(file), CompressionMode.Decompress);
        using var ms = new MemoryStream();
        gz.CopyTo(ms);
        var tar = ms.ToArray();
        var result = new Dictionary<string, byte[]>();
        var offset = 0;
        string? longName = null;
        while (offset + 512 <= tar.Length)
        {
            var header = tar.AsSpan(offset, 512);
            if (header.IndexOfAnyExcept((byte)0) < 0) break;
            string Field(int start, int length) => Encoding.UTF8.GetString(tar, offset + start, length).Split('\0')[0];
            var sizeText = Field(124, 12).Trim();
            var size = sizeText == "" ? 0 : Convert.ToInt64(sizeText, 8);
            var type = Field(156, 1);
            var prefix = Field(345, 155);
            var name = longName ?? (prefix != "" ? $"{prefix}/{Field(0, 100)}" : Field(0, 100));
            longName = null;
            var body = tar.AsSpan(offset + 512, (int)size).ToArray();
            if (type == "L") longName = Encoding.UTF8.GetString(body).TrimEnd('\0');
            else if (type == "x")
            {
                var m = Regex.Match(Encoding.UTF8.GetString(body), @"\d+ path=([^\n]+)\n");
                if (m.Success) longName = m.Groups[1].Value;
            }
            else if (type is "0" or "") result[name.StartsWith("./") ? name[2..] : name] = body;
            offset += 512 + (int)((size + 511) / 512 * 512);
        }
        return result;
    }

    // ---------------------------------------------------------------- установка

    public static DxvkState Status(string dir)
    {
        try
        {
            var data = JsonNode.Parse(File.ReadAllText(Path.Combine(dir, Marker)));
            return new DxvkState(true, data.Str("version") ?? "", data.Str("arch"), data.Str("api"));
        }
        catch { return new DxvkState(false); }
    }

    static async Task<(string Version, string File)> Archive(IProgress<InstallStep> progress, CancellationToken ct)
    {
        var data = await Http.GetJson(Releases, ct, 20, "application/vnd.github+json");
        var asset = data.Arr("assets").FirstOrDefault(a => Regex.IsMatch(a.Str("name") ?? "", @"^dxvk-[\d.]+\.tar\.gz$", RegexOptions.IgnoreCase))
            ?? throw new InvalidOperationException(I18n.T("err.DXVK_DOWNLOAD"));
        var version = (data.Str("tag_name") ?? "").TrimStart('v', 'V');
        var name = asset.Str("name")!;
        var sha = Regex.Match(asset.Str("digest") ?? "", "^sha256:([0-9a-f]{64})$", RegexOptions.IgnoreCase) is { Success: true } m ? m.Groups[1].Value : null;
        var cached = Path.Combine(Directory.CreateDirectory(Path.Combine(Paths.CacheDir, "dxvk")).FullName, name);
        if (File.Exists(cached)) return (version, cached);
        var file = await Http.Download(asset.Str("browser_download_url")!, name, new Progress<double>(r => progress.Report(new InstallStep("loader.download", $"DXVK {version}", Ratio: r))), sha, ct);
        File.Move(file, cached, true);
        return (version, cached);
    }

    public static async Task<(string Version, string Api)> Install(string exe, string? chosenApi, IProgress<InstallStep> progress, CancellationToken ct)
    {
        var game = Inspect(exe);
        var api = chosenApi is not null && Files.ContainsKey(chosenApi) ? chosenApi : game.Api;
        if (api is null || !Files.ContainsKey(api)) throw new InvalidOperationException(I18n.T("err.DXVK_API"));
        if (Status(game.Dir).Installed) Remove(game.Dir);
        var names = Files[api];
        if (names.Any(n => ReShade.IsReShadeDll(Path.Combine(game.Dir, n)))) throw new InvalidOperationException(I18n.T("err.DXVK_RESHADE"));

        var (version, archive) = await Archive(progress, ct);
        progress.Report(new InstallStep("loader.install", "DXVK"));
        var entries = ReadTarGz(archive);
        var folder = game.Arch == "x64" ? "x64" : "x32";
        byte[]? Pick(string n) => entries.FirstOrDefault(kv => kv.Key.EndsWith($"/{folder}/{n}", StringComparison.OrdinalIgnoreCase)).Value;
        var files = names.Where(n => Pick(n) is not null).ToList();
        if (files.Count == 0) throw new InvalidOperationException(I18n.T("err.DXVK_DOWNLOAD"));

        var backups = new List<string>();
        var written = new List<string>();
        try
        {
            foreach (var n in files)
            {
                var target = Path.Combine(game.Dir, n);
                if (File.Exists(target)) { File.Move(target, target + BackupSuffix, true); backups.Add(n); }
                await File.WriteAllBytesAsync(target, Pick(n)!, ct);
                written.Add(n);
            }
            var marker = new JsonObject
            {
                ["version"] = version,
                ["arch"] = game.Arch,
                ["api"] = api,
                ["files"] = new JsonArray(written.Select(x => (JsonNode)x).ToArray()),
                ["backups"] = new JsonArray(backups.Select(x => (JsonNode)x).ToArray()),
                ["exe"] = Path.GetFileName(exe),
                ["installedAt"] = DateTime.UtcNow.ToString("o"),
            };
            File.WriteAllText(Path.Combine(game.Dir, Marker), marker.ToJsonString(new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));
        }
        catch
        {
            foreach (var n in written) try { File.Delete(Path.Combine(game.Dir, n)); } catch { }
            foreach (var n in backups)
            {
                var target = Path.Combine(game.Dir, n);
                if (File.Exists(target + BackupSuffix)) File.Move(target + BackupSuffix, target, true);
            }
            throw;
        }
        progress.Report(new InstallStep("loader.done", "DXVK"));
        return (version, api);
    }

    public static bool Remove(string dir)
    {
        JsonNode? data;
        try { data = JsonNode.Parse(File.ReadAllText(Path.Combine(dir, Marker))); } catch { return false; }
        var known = Files["dx8"].Concat(Files["dx11"]).ToHashSet();
        foreach (var n in data.Arr("files").Select(x => x?.ToString() ?? ""))
            if (known.Contains(n.ToLowerInvariant())) try { File.Delete(Path.Combine(dir, n)); } catch { }
        foreach (var n in data.Arr("backups").Select(x => Path.GetFileName(x?.ToString() ?? "")))
        {
            var target = Path.Combine(dir, n);
            if (File.Exists(target + BackupSuffix)) File.Move(target + BackupSuffix, target, true);
        }
        File.Delete(Path.Combine(dir, Marker));
        return true;
    }

    // ---------------------------------------------------------------- список игр в настройках (dxvkGames)

    public sealed record Entry(string Exe, string Name, bool Exists, DxvkState State);

    public static List<Entry> List() =>
        Settings.Data.Arr("dxvkGames")
            .Where(g => g.Str("exe") is not null)
            .Select(g =>
            {
                var exe = g.Str("exe")!;
                var exists = File.Exists(exe);
                return new Entry(exe, g.Str("name") ?? Path.GetFileName(exe), exists, exists ? Status(Path.GetDirectoryName(exe)!) : new DxvkState(false));
            }).ToList();

    public static void Remember(string exe, string name)
    {
        var list = Settings.Data.Arr("dxvkGames").Where(g => g.Str("exe") != exe).Select(g => g!.DeepClone()).Take(49).ToList();
        list.Insert(0, new JsonObject { ["exe"] = exe, ["name"] = name });
        Settings.Data["dxvkGames"] = new JsonArray(list.ToArray());
        Settings.Save();
    }

    public static void Forget(string exe)
    {
        Settings.Data["dxvkGames"] = new JsonArray(Settings.Data.Arr("dxvkGames").Where(g => g.Str("exe") != exe).Select(g => g!.DeepClone()).ToArray());
        Settings.Save();
    }
}
